server/rollback_runner.go

```diff
@@ -0,0 +1,161 @@
+				return true
+			})
+		case <-ctx.Done(): // if cancel() execute
+			return
+		}
+	}
+}
+
+// retryGenerateRollbackSQL retries generating rollback SQL for tasks.
+// It is currently called when Bytebase server starts and only rerun unfinished generation.
+func (r *RollbackRunner) retryGenerateRollbackSQL(ctx context.Context) {
+	find := &api.TaskFind{
+		StatusList: &[]api.TaskStatus{api.TaskDone},
+		TypeList:   &[]api.TaskType{api.TaskDatabaseDataUpdate},
+		Payload:    "payload->>'threadID'!='' AND payload->>'rollbackError' IS NULL AND payload->>'rollbackStatement' IS NULL",
+	}
+	taskList, err := r.store.FindTask(ctx, find, true)
+	if err != nil {
+		log.Error("Failed to get running DML tasks", zap.Error(err))
+		return
+	}
+	for _, task := range taskList {
+		log.Debug("retry generate rollback SQL for task", zap.Int("ID", task.ID), zap.String("name", task.Name))
+		r.generateMap.Store(task.ID, task)
+	}
+	r.generateSignal <- struct{}{}
+}
+
+func (r *RollbackRunner) generateRollbackSQL(ctx context.Context, task *api.Task) {
+	defer func() {
+		if r := recover(); r != nil {
+			err, ok := r.(error)
+			if !ok {
+				err = errors.Errorf("%v", r)
+			}
+			log.Error("Rollback runner PANIC RECOVER", zap.Error(err), zap.Stack("panic-stack"))
+		}
+	}()
+
+	payload := &api.TaskDatabaseDataUpdatePayload{}
+	if err := json.Unmarshal([]byte(task.Payload), payload); err != nil {
+		log.Error("Invalid database data update payload", zap.Error(err))
+		return
+	}
+
+	rollbackSQL, err := r.generateRollbackSQLImpl(ctx, task, payload)
+	if err != nil {
+		log.Error("Failed to generate rollback SQL statement", zap.Error(err))
+		payload.RollbackError = err.Error()
+	}
+	payload.RollbackStatement = rollbackSQL
+
+	payloadBytes, err := json.Marshal(payload)
+	if err != nil {
+		log.Error("Failed to marshal task payload", zap.Error(err))
+		return
+	}
+	payloadString := string(payloadBytes)
+	patch := &api.TaskPatch{
+		ID:        task.ID,
+		UpdaterID: api.SystemBotID,
+		Payload:   &payloadString,
+	}
+	if _, err := r.store.PatchTask(ctx, patch); err != nil {
+		log.Error("Failed to patch task with the MySQL thread ID", zap.Int("taskID", task.ID))
+		return
+	}
+	log.Debug("Rollback SQL generation success", zap.Int("taskID", task.ID))
+}
+
+func (r *RollbackRunner) generateRollbackSQLImpl(ctx context.Context, task *api.Task, payload *api.TaskDatabaseDataUpdatePayload) (string, error) {
+	basename, seqStart, err := mysql.ParseBinlogName(payload.BinlogFileStart)
+	if err != nil {
+		return "", errors.WithMessagef(err, "invalid start binlog file name %s", payload.BinlogFileStart)
+	}
+	_, seqEnd, err := mysql.ParseBinlogName(payload.BinlogFileEnd)
+	if err != nil {
+		return "", errors.WithMessagef(err, "Invalid end binlog file name %s", payload.BinlogFileEnd)
+	}
+	binlogFileNameList := mysql.GenBinlogFileNames(basename, seqStart, seqEnd)
+
+	driver, err := getAdminDatabaseDriver(ctx, task.Instance, "", r.pgInstanceDir, r.dataDir)
```

Comment:
Do we need pgInstanceDir and r.dataDir?

Reply:
r.dataDir is needed to get the resource dir, which is in turn used by mysqlbinlog.
r. pgInstanceDir is not needed.